### PR TITLE
Support RingBuffer to get the latest sample.

### PR DESCRIPTION
**Pull request by evshary**
As a successor of https://github.com/eclipse-zenoh/zenoh/pull/782, create a RingBuffer for users to get the latest data easily.

Closes https://github.com/eclipse-zenoh/zenoh/issues/809

### CODE REVIEW

File: zenoh/tests/handler.rs
Comment (Mallets) [inline zenoh/tests/handler.rs:1]: Please add file header.

File: examples/examples/z_pull.rs [hunk 1]
```diff
@@ -31,26 +27,19 @@ async fn main() {
...
-    let arb_c = arb.clone();
-
     println!("Declaring Subscriber on '{key_expr}'...");
-    let _subscriber = session
+    let subscriber = session
         .declare_subscriber(&key_expr)
-        .callback(move |sample| {
-            arb_c.lock().unwrap().push_force(sample);
-        })
+        .with(RingBuffer::new(cache))
```
Comment (Mallets) [inline examples/examples/z_pull.rs:33]: What about renaming `cache` into `size`?

File: zenoh/src/handlers.rs [hunk 2]
```diff
@@ -15,6 +15,10 @@
 //! Callback handler trait.
 use crate::API_DATA_RECEPTION_CHANNEL_SIZE;
 
+use std::sync::{Arc, Mutex, Weak};
+use zenoh_collections::RingBuffer as RingBuffer_inner;
```
Comment (Mallets) [inline zenoh/src/handlers.rs:19]: As per CameCase naming convention, use `as RingBufferInner`.

File: zenoh/src/handlers.rs [hunk 3]
```diff
@@ -88,6 +92,54 @@ impl<T: Send + Sync + 'static> IntoHandler<'static, T>
     }
 }
 
+/// Ring buffer with a limited queue size, which allows users to keep the last N data.
+pub struct RingBuffer<T> {
+    cache: Arc<Mutex<RingBuffer_inner<T>>>,
```
Comment (Mallets) [inline zenoh/src/handlers.rs:97]: Rename `cache` to `ring`?

File: zenoh/tests/handler.rs
Comment (evshary) [reply zenoh/tests/handler.rs:1]: I copied it from the test file without the file header.
Maybe I should also add the file header to other test files.

File: zenoh/src/handlers.rs (hunk 3 above)
Comment (evshary) [reply zenoh/src/handlers.rs:97]: Sounds good

File: zenoh/src/handlers.rs (hunk 2 above)
Comment (evshary) [reply zenoh/src/handlers.rs:19]: Done

File: examples/examples/z_pull.rs (hunk 1 above)
Comment (evshary) [reply examples/examples/z_pull.rs]: It's clearer to me.

File: zenoh/tests/handler.rs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        .declare_subscriber("test/ringbuffer")
+        .with(RingBuffer::new(3))
+        .res()
+        .unwrap();
+    for i in 0..10 {
+        zenoh
+            .put("test/ringbuffer", format!("put{i}"))
+            .res()
+            .unwrap();
+    }
```
Comment (Mallets) [inline zenoh/tests/handler.rs:29]: For timing resistance I'd add a sleep of 1 second here.

Comment (evshary) [reply zenoh/tests/handler.rs:30]: Agree

**Review by Mallets — APPROVED** [review]

**Pull request merged**